src/beyond_identity/api/utils/request.rs

```diff
@@ -0,0 +1,155 @@
+use http::Method;
+use reqwest_middleware::ClientWithMiddleware as Client;
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{
+    beyond_identity::{api_token::get_beyond_identity_api_token, tenant::TenantConfig},
+    common::{config::Config, error::BiError},
+};
+
+pub async fn send_request<T, U>(
+    client: &Client,
+    config: &Config,
+    tenant_config: &TenantConfig,
+    method: Method,
+    url: &str,
+    body: Option<&T>,
+) -> Result<U, BiError>
+where
+    T: Serialize,
+    U: DeserializeOwned,
+{
+    log::debug!(
+        "Sending request: method = {:?}, url = {}, body = {:?}",
+        method,
+        url,
+        body.map(
+            |b| serde_json::to_string(b).unwrap_or_else(|_| "<serialization error>".to_string())
+        )
+        .unwrap_or_else(|| "None".to_string())
+    );
+
+    let mut request_builder = client.request(method, url).header(
+        "Authorization",
+        format!(
+            "Bearer {}",
+            get_beyond_identity_api_token(client, config, tenant_config).await?
+        ),
+    );
+
+    if let Some(body) = body {
+        request_builder = request_builder.json(body);
+    }
+
+    let response = request_builder.send().await?;
+    let status = response.status();
+    let response_text = response.text().await?;
+
+    log::debug!(
+        "Received response: status = {}, text = {}",
+        status,
+        response_text
+    );
+
+    if !status.is_success() {
+        return Err(BiError::RequestError(status, response_text));
+    }
+
+    let response_body: U = serde_json::from_str(&response_text)?;
+    Ok(response_body)
+}
+
+pub async fn send_request_paginated<T, U>(
```

Comment:
If we rewrite send_request as a middleware, this function can fully focus on the pagination logic.